source/fab/tools/tool.py

```diff
@@ -0,0 +1,192 @@
+    '''This is the base class for all tools. It stores the name of the tool,
+    the name of the executable, and provides a `run` method.
+
+    :param name: name of the tool.
+    :param exec_name: name of the executable to start.
+    :param category: the Category to which this tool belongs.
+    '''
+
+    def __init__(self, name: str, exec_name: str,
+                 category: Categories = Categories.MISC):
+        self._logger = logging.getLogger(__name__)
+        self._name = name
+        self._exec_name = exec_name
+        self._flags = Flags()
+        self._category = category
+
+        # This flag keeps track if a tool is available on the system or not.
+        # A value of `None` means that it has not been tested if a tool works
+        # or not. It will be set to the output of `check_available` when
+        # querying the `is_available` property.
+        # If `_is_available` is False, any call to `run` will immediately
+        # raise a RuntimeError. As long as it is still set to None (or True),
+        # the `run` method will work, allowing the `check_available` method
+        # to use `run` to determine if a tool is available or not.
+        self._is_available: Optional[bool] = None
+
+    @abstractmethod
+    def check_available(self) -> bool:
+        '''An abstract method to check if this tool is available in the system.
+        '''
+
+    @property
+    def is_available(self) -> bool:
+        '''Checks if the tool is available or not. It will call a tool-specific
+        function check_available to determine this, but will cache the results
+        to avoid testing a tool more than once.
+
+        :returns: whether the tool is available (i.e. installed and
+            working).
+        '''
+        if self._is_available is None:
+            self._is_available = self.check_available()
+        return self._is_available
+
+    @is_available.setter
+    def is_available(self, value: bool):
+        '''Sets a tool to be available (i.e. installed and working)
+        or not.
+        :param value: if the tool is available or not.'''
+        self._is_available = value
+
+    @property
+    def is_compiler(self) -> bool:
+        '''Returns whether this tool is a (Fortran or C) compiler or not.'''
+        return self._category.is_compiler
+
+    @property
+    def exec_name(self) -> str:
+        ''':returns: the name of the executable.'''
+        return self._exec_name
+
+    @property
+    def name(self) -> str:
+        ''':returns: the name of the tool.'''
+        return self._name
+
+    @property
+    def category(self) -> Categories:
+        ''':returns: the category of this tool.'''
+        return self._category
+
+    @property
+    def flags(self) -> Flags:
+        ''':returns: the flags to be used with this tool.'''
+        return self._flags
+
+    @property
+    def logger(self) -> logging.Logger:
+        ''':returns: a logger object for convenience.'''
+        return self._logger
```

Comment:
Yeah, not sure here. I tried to avoid the repeated import of logging etc (I really don't like code execution on import time, so all these logging calls triggered me :) ). I am not sure this is better, though it does mean none of the tools need to import/setup logging.